include/xtensor/xset_operation.hpp

```diff
@@ -0,0 +1,164 @@
+/***************************************************************************
+* Copyright (c) Johan Mabille, Sylvain Corlay and Wolf Vollprecht          *
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_XSET_OPERATION_HPP
+#define XTENSOR_XSET_OPERATION_HPP
+
+#include <algorithm>
+#include <functional>
+#include <type_traits>
+
+#include <xtl/xsequence.hpp>
+
+#include "xfunction.hpp"
+#include "xutils.hpp"
+#include "xscalar.hpp"
+#include "xstrides.hpp"
+#include "xstrided_view.hpp"
+#include "xmath.hpp"
+
+namespace xt
+{
+
+    namespace detail
+    {
+
+        template<class, typename = void>
+        struct make_lambda_isin_dispatch : std::integral_constant<int, 0> {};
+
+        template<class E>
+        struct make_lambda_isin_dispatch<E, std::enable_if_t<std::is_lvalue_reference<E>::value>>
+            : std::integral_constant<int, 1> {};
+
+        template <class E>
+        inline auto make_lambda_isin(E&& e, std::integral_constant<int, 0>)
+        {
+            return [e](const auto& t) {
+                return std::find(e.begin(), e.end(), t) != e.end(); };
+        }
+
+        template <class E>
+        inline auto make_lambda_isin(E&& e, std::integral_constant<int, 1>)
+        {
+            return [&e](const auto& t) {
+                return std::find(e.begin(), e.end(), t) != e.end(); };
+        }
```

Comment:
@JohanMabille I think you had something more elegant in mind, or not?

Reply:
What I had in mind is roughly equivalent, although a bit easier for the compiler since it only involves template specialization:

```cpp
template <bool lvalue>
struct lambda_isin
{
    template <class E>
    static auto make(E&& e)
    {
        return [&e](const auto& t) { return std::find(e.begin(), e.end(), t) != e.end(); };
    }
};

template <>
struct lambda_isin<false>
{
    template <class E>
    static auto make(E&& e)
    {
        return [e](const auto& t) { return std::find(e.begin(), e.end(), t) != e.end(); };
    }
};
```

And call it this way:

```cpp
auto lambda = detail::lambda_isin<std::is_lvalue_reference<F>::value>::make(std::forward<F>(test_elements));
```

Reply:
Better indeed, I will update